docs/proposals/Error-Messages-Improvements.md

```diff
@@ -0,0 +1,133 @@
+
+### Error messages returned by Ratify CLI commands
+
+Gina is a software engineer on the CI/CD team at Contoso, where she creates pipeline tasks incorporating Ratify CLI commands to assess artifacts according to policies. Should a policy check not pass, the corresponding artifacts are prevented from progressing in the pipeline. When policy evaluation fails, Gina sees concise, clear and actionable error messages returned by Ratify CLI commands. The error messages contain concise error descriptions, error reasons and error recommendations, allowing her to act on errors promptly.
+
+## Proposed solutions
+
+We won’t create new error message guidelines; instead, we’ll refer to the existing ones. [Azure CLI Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md#error-message) outlined a general pattern for error messages, consisting of:
+
+1. __What the error is.__
+2. __Why it happens.__
+3. __What users need to do to fix it.__
+
+The proposed improvements for Ratify error messages adhere to this general pattern and the detailed DOs and DON'Ts provided in the guidelines. The error message will also include an error code. Since Ratify already supports a list of error codes, these can be used to search for remediation in the troubleshooting guide. For example, search error code `CERT_INVALID` in [the troubleshooting guide](https://ratify.dev/docs/troubleshoot/key-management-provider/kmp-tsg#cert_invalid).
+
+The recommended format for an error message in the Ratify log is as following.
+
+```text
+"<Error Code>: <Error Description>: <Error Reason>: <Remediation>"
+```
+
+For the error messages displayed in `verifierReports`, it is recommended to add two new optional fields `errorReason` and `remediation`, which will be used when the field `isSuccess` is set to `false`:
+
+```text
+"verifierReports": [
+  {
+        "subject": "<Digest of the Artifact>"
+        "isSuccess": false,
+        "message": "<Error Description>",
+        "errorReason": "<Error Reason>",
+        "remediation": "<Remediation>"
+  }
+]
+```
+
+## Examples
+
+### Error messages displayed in the Ratify logs or returned by Ratify CLI commands
+
+For the above first example, the error message in the Ratify log can be improved to:
+
+```text
+REPOSITORY_OPERATION_FAILURE: Failed to resolve the artifact descriptor: HEAD "https://roacr.azurecr.io/v2/net-monitor/manifests/v2": GET "https://roacr.azurecr.io/oauth2/token?

```

Comment:
This example is also confusing. Why are they two HTTP requests (HEAD and GET) and how is this helping the user understand what the reason is (if this is meant to explain the reason)?

Also, having everything lowercase makes those messages really hard to read because it is not clear where a sentence ends and starts.

Reply:
@toddysm This message (HEAD and GET) was returned by ACR in this example. The ACR responded with an error and the response message was printed out. If we do not print out server response, we do not know what happened and the reason. In this case, the 401 unauthorized was also included in the message.

Reply:
Passing through errors is not a user-friendly practice. We need to catch the error from ACR and wrap it into something meaningful\ for the user of Rarify. The user of Ratify may have no knowledge of where the artifact is and even not know what artifact descriptor is. We need to provide meaningful message for the user of Ratyify and not the user of ACR.